Java compiler doesn't qualify casts in pattern destructors

Given the following in `Main.dfy`:

```dafny
module A {
  datatype Wrapper = Wrapper(val: int)
}

module B {
  datatype Wrapper = Wrapper
}

module Main {
  import opened A
  import B

  method Main() {
    var Wrapper(x) := Wrapper(0);
    print x;
  }
}
```

Trying to compile and run this to Java gives a `cannot find symbol` error because the `Wrapper` type cast isn't qualified:

```
$ dafny run -t java Main.dfy

Dafny program verifier finished with 0 verified, 0 errors
/<snip>/Main-java/Main_Compile/__default.java:15: error: reference to Wrapper is ambiguous
    java.math.BigInteger _0_x = ((Wrapper)_let_tmp_rhs0)._val;
                                  ^
  both class B_Compile.Wrapper in B_Compile and class A_Compile.Wrapper in A_Compile match
1 error
Error while compiling Java files. Process exited with exit code 1
```

<!-- This page is for Dafny bug reports and feature requests.  Please remember to include include your Dafny version number, any relevant code, and expected results. -->

<!-- Having trouble with a proof?  Consider posting your question on Stack Overflow. -->

<!-- Please do not report security issues on this public tracker.  Instead, see https://github.com/dafny-lang/dafny/blob/master/.github/SECURITY.md -->

--- a/Source/DafnyCore/Compilers/Compiler-java.cs
+++ b/Source/DafnyCore/Compilers/Compiler-java.cs
@@ -2074,7 +2074,7 @@ namespace Microsoft.Dafny.Compilers {
       if (dt is TupleTypeDecl tupleDecl) {
         return DafnyTupleClass(tupleDecl.NonGhostDims);
       }
-      var dtName = IdProtect(dt.CompileName);
+      var dtName = IdProtect(dt.FullCompileName);
       return dt.IsRecordType ? dtName : dtName + "_" + ctor.CompileName;
     }
     string DtCreateName(DatatypeCtor ctor) {

--- a/Test/git-issues/git-issue-2903.dfy
+++ b/Test/git-issues/git-issue-2903.dfy
@@ -1,0 +1,25 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %dafny /noVerify /compile:4 /spillTargetCode:2 /compileTarget:cs "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /spillTargetCode:2 /compileTarget:js "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /spillTargetCode:2 /compileTarget:go "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /spillTargetCode:2 /compileTarget:java "%s" >> "%t"
+// RUN: %dafny /noVerify /compile:4 /spillTargetCode:2 /compileTarget:py "%s" >> "%t"
+// RUN: %diff "%s.expect" "%t"
+
+module A {
+  datatype Wrapper = Wrap(val: int)
+}
+
+module B {
+  datatype Wrapper = Wrap
+}
+
+module Main {
+  import opened A
+  import B
+
+  method Main() {
+    var Wrap(x) := Wrap(0);
+    expect x == 0;
+  }
+}

--- a/Test/git-issues/git-issue-2903.dfy.expect
+++ b/Test/git-issues/git-issue-2903.dfy.expect
@@ -1,0 +1,12 @@
+
+Dafny program verifier finished with 0 verified, 0 errors
+
+Dafny program verifier did not attempt verification
+
+Dafny program verifier did not attempt verification
+
+Dafny program verifier did not attempt verification
+
+Dafny program verifier did not attempt verification
+
+Dafny program verifier did not attempt verification

--- a/docs/dev/news/2904.fix
+++ b/docs/dev/news/2904.fix
@@ -1,0 +1,1 @@
+The Dafny-to-Java compiler will now fully-qualify type casts in pattern destructors, avoiding "reference to TYPE is ambiguous" errors from javac.

